### PR TITLE
Fixes for parsing JSX tokens array

### PR DESCRIPTION
**Pull request by RReverser**
With `tokens` option enabled, esprima-fb returns incorrect array of tokens in `ast.tokens` property in following ways:
1. Duplicate `XJSText` tokens appear due to double-peeking next token into tokens array by both `expect()` in `parseXJSOpeningElement()` and `peek()` in `parseXJSElement()`.
2. Whitespaces in `<Tag> a </Tag>` were ignored and returned as pure `'a'` literal inside of `XJSText` element; this also caused children array of `<Tag> </Tag>` to be empty when `tokens` option was enabled.

Both bugs are fixed and regression test added.

Also:
1. Ignored `node_modules` folder.
2. Added running JSX tests in browser.
3. Changed versions in code to same as in package.json (required for tests).

### CODE REVIEW

File: component.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "esprima",
-    "version": "3001.0001.0000-dev-harmony-fb",
+    "version": "4001.0001.0000-dev-harmony-fb",
```
Comment (benjamn) [inline component.json]: Can you bump the 0001 part to 0002 instead?

File: esprima.js [hunk 2]
```diff
@@ -5383,7 +5383,7 @@ parseYieldExpression: true
             if (ch === '&') {
                 str += scanXJSEntity();
             } else {
-                ch = source[index++];
```
Comment (benjamn) [inline esprima.js:5386]: `ch = source[++index]` would have worked too

File: esprima.js [hunk 3]
```diff
@@ -5997,7 +5999,7 @@ parseYieldExpression: true
     }
 
     // Sync with *.json manifests.
-    exports.version = '3001.0001.0000-dev-harmony-fb';
+    exports.version = '4001.0001.0000-dev-harmony-fb';
```
Comment (benjamn) [inline esprima.js]: Same comment about 0002.

File: esprima.js (hunk 2 above)
Comment (RReverser) [inline esprima.js:5386]: Not sure, as far as I understood, `ch` should have value of `source[index]` at this point.

File: component.json (hunk 1 above)
Comment (RReverser) [inline component.json]: Sure, I just originally bumped to same as already was in `package.json`. Thought it's better not to send PR with own version, usually authors are responsible for changing it so there would be no conflict.

Comment (RReverser) [inline component.json]: @benjamn Anyway, done.

File: test/fbtest.js [hunk 4]
```diff
@@ -2043,6 +2043,152 @@ var fbTestFixture = {
...
+                    start: {line: 1, column: 0},
+                    end: {line: 1, column: 11}
+                }
+            }],
+            range: [0, 11],
+            loc: {
+                start: {line: 1, column: 0},
+                end: {line: 1, column: 11}
+            },
+            tokens: [
```
Comment (jeffmo) [inline test/fbtest.js:2117]: Kinda seems like it'd be good to have these in all tests, but I'm a little surprised that this passes given that I don't see any indication for the test runner to run with the `tokens` option enabled for this test in particular...?

I see that you added fbtest.js to the browser test runner (thanks for that, btw) but does `npm test` also work?

Comment (RReverser) [inline test/fbtest.js:2117]: > Kinda seems like it'd be good to have these in all tests, but I'm a little surprised that this passes given that I don't see any indication for the test runner to run with the tokens option enabled for this test in particular...?

Esprima passes that option by default, it just has advanced technique for comparing given test object with output which allows to pass only sub-node not entire AST (as is done in all the other test cases), but when you pass Program node, comparing it results in comparing tokens property as well.

> I see that you added fbtest.js to the browser test runner (thanks for that, btw) but does npm test also work?

Yes, sure. It works as earlier. You can pull it from my fork and try if you want.

Btw, why not to enable Travis on this fork?

Comment (jeffmo) [inline test/fbtest.js:2117]: > Btw, why not to enable Travis on this fork?

I was just thinking the same thing as I was typing that comment :)

**Comment by jeffmo** [comment]
Looks good, thanks for the diff! :)

**Pull request merged**